cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java

```diff
@@ -0,0 +1,137 @@
+    this.workflowContext = workflowContext;
+    this.action = createAction(workflowContext, instantiator, classLoader);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFactory instantiator,
+                                      ClassLoader classLoader) throws Exception {
+    Class<?> clz = Class.forName(context.getSpecification().getClassName(), true, classLoader);
+    Preconditions.checkArgument(WorkflowAction.class.isAssignableFrom(clz), "%s is not a WorkflowAction.", clz);
+    WorkflowAction action = instantiator.get(TypeToken.of((Class<? extends WorkflowAction>) clz)).create();
+    Reflections.visit(action, action.getClass(),
+                      new PropertyFieldSetter(context.getSpecification().getProperties()),
+                      new DataSetFieldSetter(context),
+                      new MetricsFieldSetter(context.getMetrics()));
+    return action;
+  }
+
+  void execute() throws Exception {
+    ClassLoader oldClassLoader = setContextCombinedClassLoader(action);
+    try {
+      initializeInTransaction();
+      runInTransaction();
+    } catch (Throwable t) {
+      LOG.error("Error executing the action {} in the Workflow {}.", workflowContext.getSpecification().getName(),
+                workflowRunId, t);
+      Throwables.propagateIfPossible(t, Exception.class);
+      throw Throwables.propagate(t);
+    } finally {
+      destroyInTransaction();
+      ClassLoaders.setContextClassLoader(oldClassLoader);
+    }
+  }
+
+  private void initializeInTransaction() throws Exception {
+    TransactionContext transactionContext = workflowContext.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+    try {
+      action.initialize(workflowContext);
+    } catch (Throwable t) {
+      LOG.error("Failed to initialize the Workflow action {} for Workflow run {}.",
+                workflowContext.getSpecification().getName(), workflowRunId);
+      transactionContext.abort(new TransactionFailureException("Transaction function failure for transaction. ", t));
+    }
+    transactionContext.finish();
+  }
+
+  private void runInTransaction() throws Exception {
+    TransactionContext transactionContext = workflowContext.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+    try {
+      action.run();
+    } catch (Throwable t) {
+      LOG.error("Failed to run the Workflow action {} for Workflow run {}.",
+                workflowContext.getSpecification().getName(), workflowRunId);
+      transactionContext.abort(new TransactionFailureException("Transaction function failure for transaction. ", t));
+    }
+    transactionContext.finish();
+  }
+
+  private void destroyInTransaction() {
+    TransactionContext transactionContext = workflowContext.getDatasetCache().newTransactionContext();
+    try {
+      transactionContext.start();
+      action.destroy();
+      transactionContext.finish();
```

Comment:
I think I commented about the code consistency. It looks different than the one in init and run. I prefer all of them have similar structure.

```java
TransactionContext txContext = workflowContext.getDatasetCache().newTransactionContext();
try {     // Only do this try-catch if it's for the destroy method.
  txContext.start();
  try {
    // Call action.initialize() or action.run() or action.destroy()
    txContext.finish();
  } catch (TransactionFailureException e) {
    txContext.abort(e);
  } catch (Throwable t) {
    txContext.abort(new TransactionFailureException("....", t));
  }
} catch (Throwable t) {
  LOG.error(....);
}
```